go/pkg/cas/upload.go

```diff
@@ -39,6 +59,358 @@ type DigestStat struct {
+//    Rewinding helps locality: there is no delay between reading the file for
+//    the first and the second times.
+//    Only one large file is processed at a time because most GCE disks are
+//    network disks. Reading many large files concurrently appears to saturate
+//    the network and slows down the progress.
+//    See also ClientConfig.LargeFileThreshold.
+func (u *uploader) visitRegularFile(ctx context.Context, absPath string, info os.FileInfo) (*repb.FileNode, error) {
+	isLarge := info.Size() >= u.LargeFileThreshold
+
+	// Lock the mutex before acquiring a semaphore to avoid hogging the latter.
+	if isLarge {
+		// Read only one large file at a time.
+		u.muLargeFile.Lock()
+		defer u.muLargeFile.Unlock()
+	}
+
+	if err := u.fsSem.Acquire(ctx, 1); err != nil {
+		return nil, err
+	}
+	defer u.fsSem.Release(1)
+
+	f, err := os.Open(absPath)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	ret := &repb.FileNode{
+		Name:         info.Name(),
+		IsExecutable: (info.Mode() & 0100) != 0,
+	}
+
+	if info.Size() <= u.SmallFileThreshold {
+		// This file is small enough to buffer it entirely.
+		contents, err := ioutil.ReadAll(f)
+		if err != nil {
+			return nil, err
+		}
+		item := uploadItemFromBlob(absPath, contents)
+		ret.Digest = item.Digest
+		return ret, u.scheduleCheck(ctx, item)
+	}
+
+	// It is a medium or large file.
+
+	// Compute the hash.
+	// TODO(nodir): reuse the buffer.
+	buf := make([]byte, u.FileIOSize)
+	dig, err := digest.NewFromReaderWithBuffer(f, buf)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to compute hash")
+	}
+	ret.Digest = dig.ToProto()
+
+	item := &uploadItem{
+		Title:  absPath,
+		Digest: ret.Digest,
+	}
+
+	if isLarge {
+		// Large files are special: locality is important - we want to re-read the
+		// file ASAP.
+		// Also we are not going to use BatchUploads anyway, so we can take
+		// advantage of ByteStream's built-in presence check.
+		// https://github.com/bazelbuild/remote-apis/blob/0cd22f7b466ced15d7803e8845d08d3e8d2c51bc/build/bazel/remote/execution/v2/remote_execution.proto#L250-L254
+
+		item.Open = func() (readSeekCloser, error) {
+			_, err := f.Seek(0, io.SeekStart)
+			return f, err
+		}
+		panic("not implemented")
+		// TODO(nodir): implement.
+	}
+
+	// Schedule a check and close the file (in defer).
+	// item.Open will reopen the file.
+
+	item.Open = func() (readSeekCloser, error) {
+		return os.Open(absPath)
+	}
+	return ret, u.scheduleCheck(ctx, item)
+}
+
+// visitDir reads a directory and its descendants, subject to the
+// predicate. The function blocks until each descendant is visited, but the
+// visitation happens concurrently, using u.eg.
+func (u *uploader) visitDir(ctx context.Context, absPath string) (*repb.DirectoryNode, error) {
+	var mu sync.Mutex
+	dir := &repb.Directory{}
+	var subErr error
+	var wg sync.WaitGroup
+
+	// This sub-function exist to avoid holding the semaphore while waiting for
+	// children.
+	err := func() error {
```

Comment:
wdyt of moving this function into its own to reduce indentation? no strong feelings.

Reply:
I considered that, but it means I'd also have to carry all the state (mutex, dir, wg, etc) via params, or maybe a separate struct, so overall it gets more complicated.

Reply:
I have a slight preference towards the many-arguments uglyness instead of 6 nesting levels uglyness. IMO it's easier to human parse the first than the second. But tbh this is pretty subjective, so I'm fine with whatever you think is best.

Reply:
I agree that 6 sounds a bit too much, but in this case moving it out to a separate top-level function would save us only one level of indentation :)